### PR TITLE
[CLOUD-2262] adding documentation about EJB remote calls on Kubernetes

### PR DESCRIPTION
**Pull request by ochaloup**
https://issues.jboss.org/browse/CLOUD-2262

Documentation for the EJB remote calls enablement on Kubernetes. It's connected with the transaction context propagation and transaction recovery.

### CODE REVIEW

**Comment by ochaloup** [comment]
@jmesnil would you be so kind and check if this is in a correct format?

File: docs/src/main/asciidoc/_developer-guide/ejb3/EJB_on_Kubernetes.adoc [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+= EJB on Kubernetes
+
+If the WildFly server is deployed on the Kubernetes then there are several
+points that you need to bear in mind when you use EJBs.
+
+NOTE: When deploying on Kubernetes you should consider the use of the link:https://github.com/wildfly/wildfly-operator[WildFly Operator].
+      It manages the Kubernetes objects in WildFly friendly way.
+      For example, it uses StatefulSet for the correct handling of EJB remoting and transaction recovery processing.
+
+The rest of this chapter assumes you run to use the link:https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/[`StatefulSet`]
```
Comment (simkam) [inline docs/src/main/asciidoc/_developer-guide/ejb3/EJB_on_Kubernetes.adoc]: `assumes you run to use` wording doesn't seem correct to me

File: docs/src/main/asciidoc/_developer-guide/ejb3/EJB_on_Kubernetes.adoc [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+== EJB calls on Kubernetes
+
+The EJB caller has two options on how to configure the remote call.
+It can be defined either as a remote outbound connection (see details at link:Admin_guide{outfilesuffix}#outbound-connections[Admin Guide, section Outbound Connections])
+or you may use a direct `InitialContext` lookup in your code.
+
+If you use either case then for the Kubernetes you need to adjust the configuration of the target node.
+For the target hostname, you need to use the DNS name of the very first pod managed by `StatefulSet`.
+
+The `StatefulSet` guarantees depend on the ordering of the pods. The pod are named by the defined order.
```
Comment (simkam) [inline docs/src/main/asciidoc/_developer-guide/ejb3/EJB_on_Kubernetes.adoc]: The pod**s** are

File: docs/src/main/asciidoc/_developer-guide/ejb3/EJB_on_Kubernetes.adoc [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+  For example in the application which is named `wildfly-server` and considering the `StatefulSet` headless service is named `wildfly-server-headless`
+  then the `http` socket binding needs to be defined in the following way
+
+  ```
+  <socket-binding name="http" port="${jboss.http.port:8080}">
+     <client-mapping destination-address="${jboss.node.name}.wildfly-server-headless"/>
+  </socket-binding>
+  ```
+
+. There is need a small workaround for remote EJB transaction recovery on Kubernetes
```
Comment (simkam) [inline docs/src/main/asciidoc/_developer-guide/ejb3/EJB_on_Kubernetes.adoc]: wording doesn't seem correct to me, I would suggest "Small workaround is needed"

File: docs/src/main/asciidoc/_developer-guide/ejb3/EJB_on_Kubernetes.adoc [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  (the issue could be tracked at link:https://issues.jboss.org/browse/WFCORE-4668[WFCORE-4668]).
+  The WildFly application server has to be configured with property `wildfly.config.url`.
+  The `wildfly.config.url` points to a XML configuration file. If we consider one being placed at `$JBOSS_HOME/standalone/configuration/wildfly-config.xml`
+  then the property is setup as `JAVA_OPTS="$JAVA_OPTS -Dwildfly.config.url=$JBOSS_HOME/standalone/configuration/wildfly-config.xml"`.
+  The `wildfly-config.xml` defines the EJB recovery authentication to be used during transaction recovery for remote EJB calls.
+  The target server has to configure a user that is permitted to receive the EJB remote calls.
+  Such a user is then configured by standard means of link:Admin_Guide{outfilesuffix}#application-realm[security configuration].
+  Let’s say there is configured a user on the target server. It's used the script ``$JBOSS_HOME/bin/add-user.sh` and the user is created under the `ApplicationRealm`.
```
Comment (simkam) [inline docs/src/main/asciidoc/_developer-guide/ejb3/EJB_on_Kubernetes.adoc]: "It’s used the script $JBOSS_HOME/bin/add-user.sh" -> "The script $JBOSS_HOME/bin/add-user.sh is used" ?

**Review by simkam — CHANGES_REQUESTED** [review]

**Comment by ochaloup** [comment]
@simkam thanks for the review!

File: docs/src/main/asciidoc/_developer-guide/ejb3/EJB_on_Kubernetes.adoc [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+Some steps are related to the server configuration, the other ones
+on your application.
+
+. Configure the WildFly cluster on Kubernetes.
+  See the link:High_Availability_Guide{outfilesuffix}#discovery-for-kubernetes[High Availability Guide, section of Kubernetes discovery].
+. All the `socket-binding` must have defined a client mapping for the DNS value mapped by `StatefulSet` headless service.
+  For example in the application which is named `wildfly-server` and considering the `StatefulSet` headless service is named `wildfly-server-headless`
+  then the `http` socket binding needs to be defined in the following way
+
+  ```
```
Comment (rhusar) [inline docs/src/main/asciidoc/_developer-guide/ejb3/EJB_on_Kubernetes.adoc]: This is markdown, not asciidoc notation.

https://asciidoctor.org/docs/asciidoc-writers-guide/#listing-and-source-code-blocks

File: docs/src/main/asciidoc/_developer-guide/ejb3/EJB_on_Kubernetes.adoc [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+  then the property is setup as `JAVA_OPTS="$JAVA_OPTS -Dwildfly.config.url=$JBOSS_HOME/standalone/configuration/wildfly-config.xml"`.
+  The `wildfly-config.xml` defines the EJB recovery authentication to be used during transaction recovery for remote EJB calls.
+  The target server has to configure a user that is permitted to receive the EJB remote calls.
+  Such a user is then configured by standard means of link:Admin_Guide{outfilesuffix}#application-realm[security configuration].
+  Let’s say there is configured a user on the target server.
+  The user is created with script `$JBOSS_HOME/bin/add-user.sh` under the `ApplicationRealm`.
+  Then the caller WildFly uses the configuration in `wildfly-config.xml` this way
+  (you may copy the content below, but replace the `>>PASTE_..._HERE<<` with user and password you configured):
+
+  ```
```
Comment (rhusar) [inline docs/src/main/asciidoc/_developer-guide/ejb3/EJB_on_Kubernetes.adoc]: This is markdown, not asciidoc notation.

https://asciidoctor.org/docs/asciidoc-writers-guide/#listing-and-source-code-blocks

**Review by rhusar — CHANGES_REQUESTED** [review]

**Comment by ochaloup** [comment]
@rhusar thanks. You are right, fixed.

Just pointing - I use it because the AsciiDoc declares compatibility with markdown https://asciidoctor.org/docs/asciidoc-syntax-quick-reference/#markdown-compatibility

**Comment by rhusar** [comment]
@ochaloup 

> Markdown compatible syntax is only available when using Asciidoctor.

e.g. GitHub doesn't seem to because ti was rendered incorrectly.

File: docs/src/main/asciidoc/_developer-guide/ejb3/EJB_on_Kubernetes.adoc [hunk 7]
```diff
@@ -0,0 +1,105 @@
+[[EJB_on_Kubernetes]]
+= EJB on Kubernetes
+
+If the WildFly server is deployed on the Kubernetes then there are several
```
Comment (jmesnil) [inline docs/src/main/asciidoc/_developer-guide/ejb3/EJB_on_Kubernetes.adoc]: "on Kubernetes"

File: docs/src/main/asciidoc/_developer-guide/ejb3/EJB_on_Kubernetes.adoc [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+The EJB caller has two options on how to configure the remote call.
+It can be defined either as a remote outbound connection (see details at link:Admin_guide{outfilesuffix}#outbound-connections[Admin Guide, section Outbound Connections])
+or you may use a direct `InitialContext` lookup in your code.
+
+If you use either case then for the Kubernetes you need to adjust the configuration of the target node.
+For the target hostname, you need to use the DNS name of the very first pod managed by `StatefulSet`.
+
+The `StatefulSet` guarantees depend on the ordering of the pods. The pods are named in the prescribed order.
+If you scale your application up to 3 replicas you may expect to name of pods like `wildfly-server-0`, `wildfly-server-1`, `wildfly-server-2`.
```
Comment (jmesnil) [inline docs/src/main/asciidoc/_developer-guide/ejb3/EJB_on_Kubernetes.adoc]: "your pods will have names such as  `wildfly-server-0`, "

File: docs/src/main/asciidoc/_developer-guide/ejb3/EJB_on_Kubernetes.adoc [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+It's expected a link:https://kubernetes.io/docs/concepts/services-networking/service/#headless-services[headless services]
+to be used along with the `StatefulSet`. With the headless service, there is ensured the DNS hostname for the pod.
+If the WildFly is run on `StatefulSet` with the headless service named `wildfly-server-headless`.
```
Comment (jmesnil) [inline docs/src/main/asciidoc/_developer-guide/ejb3/EJB_on_Kubernetes.adoc]: "If the application uses the WildFly Operator, a headless service will be created with a name such as `wildfly-server-headless`.

File: docs/src/main/asciidoc/_developer-guide/ejb3/EJB_on_Kubernetes.adoc [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+Then the DNS name of the very first pod will be `wildfly-server-0.wildfly-server-headless`.
+
+If the Wildfly is run as cluster then this use of the hostname `wildfly-server-0.wildfly-server-headless`
```
Comment (jmesnil) [inline docs/src/main/asciidoc/_developer-guide/ejb3/EJB_on_Kubernetes.adoc]: "The use of the hostname..."

**Review by jmesnil — CHANGES_REQUESTED** [review]

**Comment by ochaloup** [comment]
@jmesnil thanks!

**Comment by ochaloup** [comment]
@bstansberry @jmesnil do you think this could be merged now?

**Comment by bstansberry** [comment]
@ochaloup I see the comments from @jmesnil were addressed so I'm going to merge this.

**Pull request merged**